GC base module to identify expired contents

ajantha-bhat opened the pull request:
Introduced a `gc-base` module to identify the expired contents based on cutoff time.

The contents after or equal to cutoff time and the contents that are mapped to
commit head of live keys at the time of cutoff timestamp (for time travel) will be retained.

Two step traversal algorithm is implemented based on reflog, commit log, spark and bloom filter.
Detailed algorithm is added in javadoc of  `GCImpl#identifyExpiredContents()`

All the configuration for GC is provided from `GCParams` class.

Test cases with scenario diagrams can be found in `AbstractRestGCTest`

Note that only `IcebergTable` and `IcebergView` contents GC is supported as of now.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,69 @@
+  private static final long serialVersionUID = -1268447510930923804L;
+  private ContentKey key;
+  private final Set<Content> expiredContents = new HashSet<>();
+  private final Set<Content> liveContents = new HashSet<>();
```
Initially I thought it is good to show both live and expired contents in separate table or same table. So, I kept the Set of live contents also. 
But now I think it is just waste of storage. Any opinion on this ?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+    return mergeLiveContentResults(bloomFilterMaps);
+  }
+
+  public IdentifiedResult getIdentifiedResults(
```
This method should have a javadoc explaining what it does

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+    }
+  }
+
+  public Map<String, ContentBloomFilter> getLiveContents(
```
This method should have a javadoc explaining what it does

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+
+    List<Map<String, ContentBloomFilter>> bloomFilterMaps =
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, references.size())
```
Is `references.size()` always correct here?

ajantha-bhat replied:
it will launch `references.size()` number of tasks, finally the parallelism is based on the cores available.

snazy replied:
Is it efficient in Spark to do this? I thought this should be configurable?

ajantha-bhat replied:
executor instance are always configurable, I will open a config for this with default value `references.size()`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+
+    List<IdentifiedResult> results =
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, references.size())
```
Is `references.size()` always correct here?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+      Map<String, Instant> cutOffTimestampPerRef,
+      Instant defaultCutOffTimestamp,
+      Reference reference) {
+    NessieApiV1 api = getApi(clientBuilderConfig);
```
`api` must be closed after use

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+      Map<String, String> clientBuilderConfig,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap,
+      Reference reference) {
+    NessieApiV1 api = getApi(clientBuilderConfig);
```
`api` must be closed after use

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+      Reference reference) {
+    NessieApiV1 api = getApi(clientBuilderConfig);
+    Instant cutOffTimestamp =
+        cutOffTimestampPerRef.get(reference.getName()) == null
```
`get` with default value?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+
+  private Map<String, ContentBloomFilter> walkLiveReference(
+      NessieApiV1 api, Reference reference, Predicate<CommitMeta> liveCommitPredicate) {
+    final Logger LOG = LoggerFactory.getLogger(DistributedIdentifyUtil.class);
```
Can't this be a class-level?constant?

ajantha-bhat replied:
this happens in executor, so keeping in method level to avoid sending from driver.

snazy replied:
How would a `static` field be serialized?

ajantha-bhat replied:
ok, it seems the it can work with class-level constant. I have modified.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+      Spliterator<LogResponse.LogEntry> src = commits.spliterator();
```
The whole stream/spliterator logic here is quite compex - should be moved elsewhere and have unit-test coverage.

ajantha-bhat replied:
coverage report says it is covered with functional tests. Still unit test is required ?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+    final Logger LOG = LoggerFactory.getLogger(DistributedIdentifyUtil.class);
+    Map<String, ContentBloomFilter> bloomFilterMap = new HashMap<>();
+    Set<ContentKey> deletedKeys = new HashSet<>();
+    Consumer<LogResponse.LogEntry> perCommit =
```
Does this need to be declared here? It's confusing to declare a `Consumer` here, just delegate to another function but use it 40 lines below.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+          .getOperations()
+          .forEach(
+              operation -> {
+                if (operation instanceof Operation.Delete) {
```
I suspect, this doesn't handle the following case (commits in chronological order):

--> cutoff-time is here
1. create table "foo"
2. rename table "foo" to "bar"
3. create table "foo"
4. drop table "foo"

Feels like content put to the "first foo" before it's being renamed to "bar" will not be added to the bloom filter, right?

ajantha-bhat replied:
yes, you are right. I thought of having a distinction between rename table and delete table and skip adding keys to drop table if it is rename. But forgot to implement! 

good catch, I will add it.

ajantha-bhat replied:
As I traverse from the head and identify live, from second commit I can check if the keys are deleted and content is not previously marked live using bloom filter (if it is marked live, it cannot be the drop table operations). 

This way I wont add the rename table contents into expired contents.

ajantha-bhat replied:
had to use 100% brain for this solution 🤯

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+                  Content content = ((Operation.Put) operation).getContent();
+                  boolean isExpired;
+                  ContentBloomFilter bloomFilter = liveContentsBloomFilterMap.get(content.getId());
+                  if (bloomFilter == null) {
```
Better rewrite to a ternary and move the decl of isExpired here

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,301 @@
+import org.projectnessie.model.Content;
+import org.projectnessie.model.LogResponse.LogEntry;
+
+public abstract class AbstractRestGCTest extends AbstractRestGC {
```
There should definitely be tests that illustrate and validate the case of bloom-filter-false-positives - probably requires some mocking.

ajantha-bhat replied:
mockito didn't work with quarkus test last time, I will try if any other options

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,71 @@
+
+  public ContentBloomFilter() {
+    // TODO: make it lazy initialization based on content
+    this.icebergContentFilter = BloomFilter.create(Funnels.longFunnel(), SIZE);
```
There should be a way to
1. determine the size of the bloom filter
2. specify the fpp

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, references.size())
+            .map(
+                reference ->
```
Does Spark handle lambdas now?

ajantha-bhat replied:
RDD supports lambdas from the beginning version itself I guess.

snazy replied:
Pinging @rymurr on this, as I recall there was something around lambdas in the very first GC implementation.

snazy reviewed (changes requested):
Did a first review-round.
The approach in general looks good, but there are a couple of things that need to be verified and simplified.

The changes to `AbstractRest*` that introduce another test-class hierarchy and add new test-classes (only for Quarkus??) can be simplified and re-use the existing REST-test hierarchy.

Regarding Spark, I'm a bit concerned about the use of lambdas - did that change in Spark recently?

The GC algo itself might behave wrong in some scenarios.

snazy commented:
Forgot to mention: I suspect, GC also needs to handle the new Iceberg views.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,270 @@
+import org.projectnessie.versioned.persist.gc.IdentifiedResult;
+
+/** See {@link AbstractTestRest} for details about and reason for the inheritance model. */
+public abstract class AbstractRestGC extends AbstractRest {
```
a. I don't want to run GC for all the backend store. As there is nothing specific to test for backend store and GC is slow operation
b. I need to restart quarkus server just before running the GC tests, else it works on all the dead references and makes it slow. 

Hence the separate hierarchy. Regarding few duplicate code, I will see if I can extract and use as a common code.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,270 @@
+    @Override
+    public int compare(Content c1, Content c2) {
+      if (c1.getType().equals(Content.Type.ICEBERG_TABLE)) {
+        return Long.compare(
```
caller is populating the expected and actual list per content id only. So, when it comes here content id will be same

snazy replied:
Why does it compare the "id" for other content-types then?

ajantha-bhat replied:
states that have global state should skip comparing entire content object as the contents fetched from log will have latest global state.
So, iceberg content and view content compares only snapshot id and version respectively.

delta lake content compares everything. (which also includes content id and which will be same anyways)

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,278 @@
+                    computeExpiredContents(
+                        clientBuilderConfiguration, liveContentsBloomFilterMap, reference))
+            .collect();
+    // TODO: write each task output directly into the spark table.
```
Each executor task can return an iterator of spark row and which can be written directly to the spark table. Instead of accumulating/merging each task output to a single in-memory map.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  private void setIfAbsent(ContentKey key) {
+    if (this.key == null) {
```
key is the content key (db + table name). Reference may not matter to the consumer as it is Nessie concept. consumer might care about what what table these contents belong. So that cleanup operations can be called on that key.

snazy replied:
But what does the table name mean without a reference name?

ajantha-bhat replied:
For example, Iceberg expire snapshot is a consumer of GC results. It doesn't care what reference table belongs. It just needs the table name to call the expire snapshots. Hence not keeping the reference name in the output.

snazy replied:
And what happens when the table's renamed? Or when completely different tables tables have the same name in different references? Or when a table's being recreated with the same name?

ajantha-bhat replied:
keeping reference name still may not help the consumer of GC in this situation if we don't have content id to reference mapping (including the dropped tables).
So we need to introduce this mapping? Any suggestions?

ajantha-bhat replied:
**Agree, we need reference info also**. 
Initially I was thinking as the global state is common, only once I can call expire snapshot per content id (table) on default reference.
But the table can be renamed or may not exist in default reference. So, I have to call expire snapshots for table@reference. Internally I have to skip passing the reference which are cleaned up already by checking against global state's snapshots list.

ajantha-bhat replied:
And I cannot merge the snapshot ids for per content id. I have to keep separately per reference. 
I need to change
`content id -> content key + list of all the snapshots from all the reference.`
to 
`refName -> content id -> content key +  list of expired snapshots of this reference.`

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,71 @@
+  public boolean put(Content content) {
+    if (content.getType().equals(Content.Type.ICEBERG_TABLE)) {
+      return icebergContentFilter.put(((IcebergTable) content).getSnapshotId());
+    } else if (content.getType().equals(Content.Type.DELTA_LAKE_TABLE)) {
```
yep. views too. IIRC last time I got comment that GC should be generic.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,69 @@
+
+  private static final long serialVersionUID = -1268447510930923804L;
+  private ContentKey key;
+  private final Set<Content> expiredContents = new HashSet<>();
```
Actually As GC is generic for other contents also, keeping it like this.
Later if we require GC for specific content type, we can have its ContentValues class.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,69 @@
+  private final Set<Content> liveContents = new HashSet<>();
+
+  public ContentKey getKey() {
+    return key;
```
will be used by the consumer of the GC output , key is needed for the consumer to call the cleanup procedures.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,125 @@
+              } else if (entry.getOperation().equals("ASSIGN_REFERENCE")) {
+                hash = entry.getSourceHashes().get(0);
+              } else {
+                throw new RuntimeException(
```
yeah, that means something is wrong in the system.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,125 @@
+  private List<Reference> collectDeadReferences(NessieApiV1 api) {
+    List<RefLogResponse.RefLogResponseEntry> reflogResponseEntries;
+    try {
+      reflogResponseEntries =
```
Any suggestions ? I couldn't find a better way at the moment.

ajantha-bhat replied:
we might have to periodically clean up reflog too then.
Let's see what @rymurr thinks.

I even have a crazy idea of adding a marker row in reflog for every 250 entries. So that I can query the marker rows first using database filter query and then distributively reflog traversal per between marker row reflog ids for faster results. But may be too much for now.

ajantha-bhat commented:
> The changes to AbstractRest* that introduce another test-class hierarchy and add new test-classes (only for Quarkus??) can be simplified and re-use the existing REST-test hierarchy.

a. I don't want to run GC for all the backend store. As there is nothing specific to test for backend store and GC is slow operation
b. I need to restart quarkus server just before running the GC tests, else it works on all the dead references and makes it slow.
Hence the separate hierarchy. Regarding few duplicate code, moved to common code class

>  Regarding Spark, I'm a bit concerned about the use of lambdas - did that change in Spark recently?

I think RDD support lambdas, only thing is that arguments should be serializable. I carefully passing only config and building the objects on executor itself.

> The GC algo itself might behave wrong in some scenarios.

I think you meant rename scenario. handled it now.

> Forgot to mention: I suspect, GC also needs to handle the new Iceberg views.

Handled now.

**@snazy: I have addressed all the comments, please have look at it again.**

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,125 @@
+                  "reflog.operation == 'DELETE_REFERENCE' || reflog.operation == "
+                      + "'ASSIGN_REFERENCE'",
+                  OptionalInt.empty())
+              .collect(Collectors.toList());
```
Still, the `collect()` here is not necessary - just superfluous to .stream() on the .collect() here, right?

ajantha-bhat replied:
yeah, I mapped this comment to below `collect()` in the same function and got confused. No need of `collect()` at this place.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,130 @@
+        .map(
+            entry -> {
+              String hash;
+              if (entry.getOperation().equals("DELETE_REFERENCE")) {
```
Migrate this if-chain to `switch`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,130 @@
+                throw new RuntimeException(
+                    entry.getOperation() + " operation found in dead reflog query");
+              }
+              if (entry.getRefType().equals("Branch")) {
```
Migrate this if-chain to `switch`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,130 @@
+                throw new RuntimeException(
+                    entry.getOperation() + " operation found in dead reflog query");
+              }
+              if (entry.getRefType().equals("Branch")) {
```
Actually, this code construct should reuse functionality in `RefUtil`

ajantha-bhat replied:
for that GC need to depend on one more module, so added constant in `RefLogResponseEntry`

snazy commented:
> I have addressed all the comments, please have look at it again.

Not all of them AFAICS

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,71 @@
+  private static final long serialVersionUID = -693336833916979221L;
+
+  private final BloomFilter<Long> icebergContentFilter;
+  private final BloomFilter<String> deltaLakeContentFilter;
```
Instead of big binary bloom filter, I am keeping content specific bloom filters (int type for views, long type for IcebergTable, string type for deltalake)

snazy replied:
I get that. But why do you need two at all?

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,71 @@
+      return icebergContentFilter.put(((IcebergTable) content).getSnapshotId());
+    } else if (content.getType().equals(Content.Type.DELTA_LAKE_TABLE)) {
+      // TODO: understand delta lake content object and handle accordingly
+      return deltaLakeContentFilter.put(((DeltaLakeTable) content).getLastCheckpoint());
```
you are suggesting to throw unsupported exception for this content type ?

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,71 @@
+      // TODO: understand delta lake content object and handle accordingly
+      return deltaLakeContentFilter.put(((DeltaLakeTable) content).getLastCheckpoint());
+    } else {
+      // TODO: throw exception for the unsupported types?
```
Ideally this will never happen. Just a safe guard code to catch new content type impl.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,270 @@
+      throws NessieNotFoundException, NessieConflictException {
+    IcebergTable meta =
+        IcebergTable.of(
+            prefix + "_" + metadataFile, snapshotId, 42, 42, 42, prefix + "_" + contentId);
```
For Iceberg GC, only snapshot id is needed for the output for cleanup. It doesn't matter whether the how many content with same snapshot id exist. So, not caring about other on-ref params here.

ajantha-bhat commented on model/src/main/java/org/projectnessie/model/Reference.java:
```diff
@@ -50,7 +51,7 @@
 @JsonSubTypes({@Type(Branch.class), @Type(Tag.class), @Type(Detached.class)})
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
-public interface Reference extends Base {
+public interface Reference extends Base, Serializable {
```
Sorry, I didn't get this comment. What I should do ?

snazy replied:
In some places you add a `serialVerionUid` but not in others. Since the classes are only serialized, and only guaranteed to be serializable w/o conflicts - aka no backwards compatiblity, it's not necessary to add `serialVerionUid`s.

ajantha-bhat commented:
> Not all of them AFAICS

@snazy : I think this time I handled all. I have resolved the comments now which was surely handled. For the comments that I am not sure, I have replied a comment why it is open. Please reply to that if it is still not resolved else you can make it resolved so I will get the idea.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+    }
+  }
+
+  private void handleCommit(
```
Should the method name be "handleCommitForExpiredContents"  or so? same applies to it's call site

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+        private boolean refHead = true;
+
+        @Override
+        public boolean tryAdvance(Consumer<? super LogResponse.LogEntry> action) {
```
This Spliterator should really be unit tested and live in a separate utility class. It's too easy to miss cases here, that are afterwards hard to detect / debug.

ajantha-bhat replied:
done. moved to a Util class and added a UT.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+          reference,
+          commitMeta ->
+              // If the commit time is newer than (think: greater than or equal to) cutoff-time,
+              // then
```
(cah be a single comment line)

ajantha-bhat replied:
yeah, spotless is not that smart I guess ;)

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+      return walkLiveReference(
+          api,
+          reference,
+          commitMeta ->
```
This predicate is missing a very important check:
All _young_ commits _must_ be considered live. So all commits that happened within e.g. the last 2-3 hours are young - that time should be configurable, but a default of 2h feels fine.
This must happen to ensure that GC does not accidentally consider a freshly written commit "expired"

ajantha-bhat replied:
Agree. As it is a two step process, there can be new commits when the step two queries all the commits. 
So, it is good to have configurable `nessie.gc.commit.protection.time.in.hours` (default 2 hours, even though step 1 will ideally finish faster).

I think using it in step 2 is enough. Because If I never consider commits for expiry check in step 2, they can never be expired.

ajantha-bhat replied:
Handled.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+    List<IdentifiedResult> results =
+        new JavaSparkContext(session.sparkContext())
+            .parallelize(references, getTaskCount(references))
+            .map(reference -> computeExpiredContents(config, liveContentsBloomFilterMap, reference))
```
No need to pass the instance field `config` to the instance method `computeExpirecContents`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+            .map(
+                reference ->
+                    computeLiveContents(
+                        config, cutOffTimestampPerRef, defaultCutOffTimestamp, reference))
```
No need to pass the instance field `config` to the instance method `computeExpirecContents`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+            .parallelize(references, getTaskCount(references))
+            .map(
+                reference ->
+                    computeLiveContents(
```
Compute the cut-off-time for `reference` here - so don't serialize the whole `cutOffTimestampPerRef` + `defaultCutOffTimestamp`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,134 @@
+ * Encapsulates the logic to retrieve all content including their content keys over all commits in
+ * all named-references.
+ */
+public class GCImpl {
```
One area that feels worth to be explored to save some work when traversing all references:
When you retrieve all references, you can get the common-ancestor with the default branch.
Assuming that the cut-off-time-interval for a reference is less-than-or-equal to the one of the default branch, you could actually stop traversing reference after processing the common ancestor commit from all references.

That's just an idea, not sure whether that works in practice - so nothing for this PR though - but probably worth a follow-up issue.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java:
```diff
@@ -0,0 +1,134 @@
+            entry -> {
+              String hash;
+              switch (entry.getOperation()) {
+                case "DELETE_REFERENCE":
```
Use constants for these

snazy commented on model/src/main/java/org/projectnessie/model/RefLogResponse.java:
```diff
@@ -41,6 +41,10 @@
       return ImmutableRefLogResponseEntry.builder();
     }
 
+    // maps to getRefType() output
```
Add a note that the values must correspond to the enum names in `persist.proto`

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,246 @@
+        .getLogEntries();
+  }
+
+  void fillExpectedContents(Branch branch, int limit, Map<String, Set<Content>> expected)
```
`limit` should be `numCommits`

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,246 @@
+public abstract class AbstractRestGC extends AbstractRest {
+
+  @NotNull
+  List<LogEntry> fetchLogEntries(Branch branch, int limit) throws NessieNotFoundException {
```
`limit` should be `numCommits`

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,246 @@
+    List<LogEntry> logEntries = fetchLogEntries(branch, limit);
+    logEntries.stream()
+        .map(LogEntry::getOperations)
+        .forEach(
```
Replace `.forEach` with `flatMap` (it's more readable)

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,246 @@
+  }
+
+  private void verify(
+      IdentifiedResult contentValuesPerType,
```
`contentValuesPerType` : which type??

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,246 @@
+      Map<String, Set<Content>> expectedLive,
+      Map<String, Set<Content>> expectedExpired,
+      Set<String> expectedContentIds) {
+    AssertionsForClassTypes.assertThat(contentValuesPerType.getContentValues()).isNotNull();
```
Prefer static imports of all the AssertJ methods

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,416 @@
+
+  @Test
+  public void testSingleRefMultiTable() throws BaseNessieClientServerException {
+    String prefix = "test1";
```
Those should really be the name of the test method - not a number (for all tests)

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,416 @@
+            "file3",
+            b2table1.content,
+            null);
+    fillExpectedContents(branch1, 1, expectedExpired);
```
Why is this expired? It's committed after cut-off-time and branch1 is still alive?

ajantha-bhat replied:
because I am dropping the table "dropping this table" at line 301.

Yeah, bit of puzzle to understand. we have to go through full testcase, cannot understand from the current line. Let me add comments then

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,416 @@
+    table2 =
+        commitSingleOp(
+            prefix, branch2, table2.hash, 44, "cid_2", "table_2", "file3", table2.content, null);
+    fillExpectedContents(branch2, 1, expectedExpired);
```
Why is this expired? It's committed after cut-off-time and branch2 is still alive?

ajantha-bhat replied:
this is per reference cutoff time testcase, for branch2 cutoff time is after this commit (line 347)

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,416 @@
+    table2 =
+        commitSingleOp(
+            prefix, branch2, table2.hash, 44, "cid_2", "table_2", "file3", table2.content, null);
+    fillExpectedContents(branch2, 1, expectedExpired);
```
Why is this expired? It's committed after cut-off-time and branch2 is still alive?

ajantha-bhat replied:
because it is assign reference (dead commit). Line 401

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,416 @@
+            prefix, branch, table2.hash, 44, "cid_2", "table_2", "file3", table2.content, null);
+    // drop table "table_2"
+    dropTableCommit(prefix, branch, table2.hash, "table_2");
+    fillExpectedContents(branch, 2, expectedExpired);
```
Why is this expired? It's committed after cut-off-time and table2.hash is still a live commit, right?

ajantha-bhat replied:
dropping this table in line 119 and there is only one reference. so it is not live commit.

snazy reviewed (changes requested):
Looks much better now.

But I suspect there's a bug with handling expired data (see comments in `AbstractRestGCTest`).

snazy commented:
Don't forget to put `0.19.1-SNAPSHOT` into the new pom.xml files

ajantha-bhat commented on servers/quarkus-server/src/test/java/org/projectnessie/server/gc/TestRestApiGcInMemory.java:
```diff
@@ -0,0 +1,25 @@
+
+@QuarkusTest
+@TestProfile(QuarkusTestProfileInmemoryGC.class)
+class TestRestApiGcInMemory extends AbstractRestGCTest {}
```
ok. Moved.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRest.java:
```diff
@@ -31,7 +41,66 @@
+  private NessieApiV1 api;
+  private HttpClient httpClient;
+
+  static {
```
It is not huge change, so I am planning to keep it inside this as it is more understandable with this PR.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,246 @@
+    spark.sparkContext().setLogLevel("WARN");
+    try {
+      final Map<String, String> options = new HashMap<>();
+      options.put(CONF_NESSIE_URI, getHttpClient().getBaseUri().toString());
```
ok

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java:
```diff
@@ -0,0 +1,309 @@
+   * @param defaultCutOffTimestamp default cutoff timestamp to be used.
+   * @return map of {@link ContentBloomFilter} per content-id.
+   */
+  public Map<String, ContentBloomFilter> getLiveContentsBloomFilters(
```
we don't have ContentId class and ContentId is string type, so returning string.

Do you mean contentKey? Each ref can call table in its own name. So, returning  ContentId instead of contentKey.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,64 @@
+
+  private static final long serialVersionUID = -1268447510930923804L;
+  private ContentKey contentKey;
+  private final Set<Content> expiredContents = new HashSet<>();
```
That was my question also when I opened the PR. I will just keep the expired contents then

![image](https://user-images.githubusercontent.com/5889404/152956251-877c27b7-3310-4248-b9ec-53d843628bc5.png)

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,133 @@
+  // track iceberg contents only using the snapshot id (long bloom filter)
+  // as the consumer of GC results needs only this info for clean up.
+  private volatile BloomFilter<Long> icebergContentFilter;
+  // TODO: understand delta lake content object and handle accordingly
```
IIRC @rymurr wanted a generic base gc module that can handle delta lake also.
Now also I have implemented generic framework but bloom filter is content type specific. 

As we don't have plans for delta lake GC consumer implementation, I will remove it as you suggested. In future it is very small task to plugin the delta lake GC to this framework.

ajantha-bhat commented:
> But I suspect there's a bug with handling expired data (see comments in AbstractRestGCTest).

@snazy: I have gone through the comments, I don't think there is an issue for handling expired data. I have replied in the comments. But I will add comments for testcase now. so that it will be easy to know why.

still 30% of comments are pending to address (unit test for spliterator, guard railing the young commits, keeping only expired contents in the output). I will handle today and ping here again. Thanks.

ajantha-bhat commented:
@snazy : Addressed all the comments.

But what and all need to be stored in `IdentifiedResults` (output of GC) need to be finalized. It will be small change but big decision. Will conclude it today.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,236 @@
+  static class ContentComparator implements Comparator<Content> {
+    @Override
+    public int compare(Content c1, Content c2) {
+      if (c1.getType().equals(Content.Type.ICEBERG_TABLE)) {
```
Should use `switch (c1.getType()` and also consider Iceberg views.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,104 @@
+    return viewContentFilter;
+  }
+
+  public boolean put(Content content) {
```
Nit: Return value's never used.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,47 @@
+
+  private static final long serialVersionUID = -1268447510930923804L;
+  private ContentKey contentKey;
+  private final Set<Content> expiredContents = new HashSet<>();
```
I still think that `Content` should be replaced here with something like this:

```java 

  interface GcContentIdent {
    String getId();
  }
  @Value.Immutable
  interface IcebergTableGcIdent extends GcContentIdent {
    long getSnapshotId();
    static IcebergTableGcIdent fromContent(Content content) {
      ...
    }
  }
  @Value.Immutable
  interface IcebergViewGcIdent extends GcContentIdent {
    int getVersionId();
    static IcebergViewGcIdent fromContent(Content content) {
      ...
    }
  }
```

Currently, both `IcebergTable` and `IcebergView` are rather small objects and the "win" using the above pretty low.
But once we store the whole table/view-metadata in Nessie, those objects will become pretty big.

I'd be happy, if you can do this refactor in a follow-up.

ajantha-bhat replied:
I want to keep it generic now, based on the need and advantages, I can do a follow up.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,306 @@
+          commitMeta ->
+              // If the commit time is newer than (think: greater than or equal to) cutoff-time,
+              // then commit is live.
+              Objects.requireNonNull(commitMeta.getCommitTime()).compareTo(cutOffTimestamp) >= 0);
```
Nit: The usage of `Objects.requireNonNull` here adds no value. Either add a descriptiive message or just omit the `Objects.requireNonNull`

Same for the user usages of `Objects.requireNonNull` below

ajantha-bhat replied:
I added for IDE warning. I will remove it if not required.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGC.java:
```diff
@@ -0,0 +1,236 @@
+      Instant cutoffTimeStamp,
+      Map<String, Instant> cutOffTimeStampPerRef,
+      Map<String, Set<Content>> expectedExpired,
+      Set<String> expectedContentIds,
```
Nit: you can make the actual tests more readable by migrating the method signature to
```java 
void performGc(
      Instant cutoffTimeStamp,
      Map<String, Instant> cutOffTimeStampPerRef,
      Map<String, Set<Content>> expectedExpired,
      boolean disableCommitProtection,
      String... expectedContentIds)
```
and construct the `Set` from `expectedContentIds` here. Maybe even more, if you pass the `prefix` into this method as well.

OTOH do you need `expectedContentIds` at all, because those should already be in `expectedExpired`, no?

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,427 @@
+            prefix, branch1, table2.hash, 44, "cid_2", "table_2", "file3", table2.content, null);
+    // drop table "table_2"
+    dropTableCommit(prefix, branch1, table2.hash, "table_2");
+    // table_2's one commit (after cutoff timestamp) is also expected to be expired as table is
```
I suspect, this is wrong.
The commit created in line 124 is after the cut-off time and should be considered "live" / usable via time-travel.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,427 @@
+
+    // rename table "table_2" to "table_2_renamed" on branch1
+    table2 =
+        commitSingleOp(
```
This will result in `cid_2` being accessible as `table_2` *and* `table_2_renamed`.
Aka: the commit for the table-rename here must include a `Delete` operation for `table_2` and a `Put` operation for `table_2_renamed`

ajantha-bhat replied:
It is handled already. I am passing `beforeRename` variable and based on that it is `put` and `delete` inside. other cases it is just `put`.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,427 @@
+    table2 =
+        commitSingleOp(
+            prefix, branch2, table2.hash, 44, "cid_2", "table_2", "file3", table2.content, null);
+    // expect this commit to be expired as branch2 is dropped below and there is no other reference
```
This sounds legit and correct. What I'm not sure about is whether user shall have the ability to "undo a drop of a branch/tag", which would only work until GC ran.
OTOH this should be protected via the "new commit protection", right?

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,427 @@
+            "file3",
+            b2table1.content,
+            null);
+    // expect this commit (after cutoff timestamp) to be expired as table is dropped below on
```
I suspect, this is wrong.
The commit created here is after the cut-off time and should be considered "live" / usable via time-travel.
Same for line 305 below

ajantha-bhat replied:
I got a comment for showing dropped table's content irrespective of commit time to be expired on my old PR from @rymurr. Hence I added this. This looks like a conflicting comment and we need to discuss and conclude on this

ajantha-bhat replied:
we should support time travel for live tables only ?

snazy replied:
The table *is* live, because it is visible in a commit *after* cut-off time.

ajantha-bhat replied:
but it is dropped after cut-off time. This table is alive in other reference. So, the commits on this table made after cut-off time in other reference is still kept. Only commits of this table on this reference is expired as this particular content is not present in any of the reference.

snazy reviewed (changes requested):
Still think that GC's too aggressive here - IMO is should retain all contents on all references after cut-off time, because those are considered "live" / reachable via time-travel.

Regarding your question how to actually run Iceberg's expire-snapshot:
That can IMO only work properly, if you run expire-snapshot on a temporary branch created by GC. On that branch, add Put operations for all contents that need to be expired (the table name doesn't matter on that temporary branch). Then let expire-snapshot work on that branch.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,104 @@
+    this.fpp =
+        Double.parseDouble(
+            config.getOrDefault(
+                "nessie.gc.allowed.fpp.per.table.bloomfilter", String.valueOf(0.03D)));
```
It is the same default value in guava's bloom filter. So, I selected this

snazy commented:
> > Every content that is visible in any commit after cut-off time must survive - no matter what.
> 
> @snazy : So, even if the **reference** is dropped **after cutoff time**, we need to keep those contents as live?
> 
> From your drop table example, I can understand what is needed. But I cannot understand why it is needed.

Every reachable content after cut-off time MUST be accessible to allow time travel

ajantha-bhat commented:
> Still think that GC's too aggressive here - IMO is should retain all contents on all references after cut-off time, because those are considered "live" / reachable via time-travel.

@snazy : Fixed it. Also found a bug in algorithm regarding multiple keys scenario. Fixed that also.
updated the test and UT. Also updated the algorithm description.

Please have a look. Thanks.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,104 @@
+    this.expectedEntries =
+        Long.parseLong(
+            config.getOrDefault(
+                "nessie.gc.expected.per.table.bloomfilter.size", String.valueOf(1000000)));
```
added all the constants in [NessieConfigConstants.java]

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,104 @@
+    this.expectedEntries =
+        Long.parseLong(
+            config.getOrDefault(
+                "nessie.gc.expected.per.table.bloomfilter.size", String.valueOf(1000000)));
```
yes, Initialized as you suggested.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,370 @@
+    // one commit for "table_1" on branch1
+    CommitOutput table1 =
+        commitSingleOp(
+            prefix, branch1, branch1.getHash(), 42, "cid_1", "table_1", "file1", null, null);
```
Can you add constants for all these strings in all the tests?
That makes it easier to follow what's going on in IntelliJ.

ajantha-bhat replied:
added

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,370 @@
+    Map<String, Map<String, Set<Content>>> expectedExpired = new HashMap<>();
+
+    Branch before = createBranch(prefix + "_0");
+    Branch branch1 = createBranch(prefix + "_1");
```
Lines 333-359 and 162-187 are identical - so are the test cases actually. So worth to either parameterize or extract the common parts and move those tests next to each other.

ajantha-bhat replied:
merged the test cases as these are after cutoff time operations and doesn't really impact.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,370 @@
+            prefix, branch1, table2.hash, 43, "cid_2", "table_2", "file2", table2.content, null);
+    // All 3 commits are expected to be expired as it is before the cutoff time and not the head
+    // commits.
+    fillExpectedContents(branch1, 4, expectedExpired);
```
The comment says *3* commits, but here there are *4* commits. That's confusing.

ajantha-bhat replied:
it should be 3 commits in the arguments. But there are totally 3 commits itself. So no functional impact. I will fix it.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,370 @@
+            prefix, branch1, branch1.getHash(), 42, "cid_2", "table_2", "file1", null, null);
+    table2 =
+        commitSingleOp(
+            prefix, branch1, table2.hash, 43, "cid_2", "table_2", "file2", table2.content, null);
```
It's a bit tricky to realize, but I'm certain that this commit must survive.
cutoff-time is before the commit around line 60, so a time-travel to any time between `cutoffTime` and the commit-time of the commit around line 60, must return table-2@snapshot-43.

You cannot use the 1st commit after `cutoffTime`, but the 1st commit _before_ `cutoffTime`.

The 1st commit after `cutoffTime` can never happen or happen after a long time.

ajantha-bhat replied:
oh no.. Algorithm becomes too conditional and messy with time travel :(

ajantha-bhat replied:
Add all the contents after or equal to cutoff timestamp to the live contents. Then get the live keys at cutoff timestamp. Then for each live key keep the head commits as live for commits after cutoff timestamp. This should work. I will modify.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,370 @@
+    // All 4 commits are expected to be expired as it is before the cutoff time and there are some
+    // more commits for
+    // each table after cutoff timestamp (i,e these 4 commits are not head commits)
+    fillExpectedContents(branch1, 4, expectedExpired);
```
Same bug as above.

ajantha-bhat replied:
fixed it and in one more testcase.

snazy reviewed (changes requested):
Sorry, but there's still a major issue here.
You cannot just GC "everything" before cutoff-time - the first commit *before* cutoff-time is the state that must be retained.

ajantha-bhat commented:
> Sorry, but there's still a major issue here.
You cannot just GC "everything" before cutoff-time - the first commit before cutoff-time is the state that must be retained.

@snazy : Fixed all the comments. Thanks for helping me understanding the scenarios better.

ajantha-bhat commented:
@snazy : 
oh, I found rename table is not time travel friendly. Now I rename a table and when I go sometime back and do get entries with hash on reference, I get no entry. Instead of keys with old table name. Let me check this.

ajantha-bhat commented:
the scenario is bloom filter is formed only for live ref, so bug with dropped reference scenario.
bloom filter creation need to support traversing all the reference not just live reference. So that dropped reference's before cutoff time contents are maintained for time travel.
will fix and push now.

ajantha-bhat commented:
@snazy : Please have a look at it again, fixed the time travel scenario for dropped ref also. Test cases are updated with some diagrams for easy understanding.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,482 @@
+        .assignTo(before)
+        .assign();
+
+    // drop ref branch1 should not affect as it is performed after cutoff timestamp.
```
I suspect there should also be a config for the cutoff-time for content on dead references.

ajantha-bhat replied:
the current configuration works for dead references also including the commit protection time config.

snazy replied:
Dead references might not be expected to be retained for the duration of the default cutoff-time, which can be quite long. So a separate config makes sense.

ajantha-bhat replied:
we anyway support per reference configuration. We can configure cutoff time for dead references using that ?

ajantha-bhat replied:
we can also have it in a follow up if it is really required.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,108 @@
+  private volatile BloomFilter<Long> icebergTableFilter;
+  // track view contents only using the version-id (int bloom filter)
+  // as the consumer of GC results needs only this info for clean up.
+  private volatile BloomFilter<Integer> icebergViewFilter;
```
The implementation only needs one BF, that uses the content-type+id. Having one BF per content type is not necessary, but wastes a lot of space. It also complicates this class with all the `if-else` chains.

ajantha-bhat replied:
Sorry, I expect this design to be scalable when we add more content types like delta lake.
Having different bloom filter per content type will make sense for it. It may need binary filter or string filter.

ajantha-bhat replied:
Also it is lazy initialization. If Iceberg view content type is not present for that reference, then that bloom filter will not be created.

snazy replied:
How is it "scaleable" when you have to touch this code and add even more if-else clauses, allocate more heap space etc etc?
A BF does not care about the "source type".
And further: the presence of a single content allocates memory for a BF for the total number of commits, which is a huge waste of resources.

ajantha-bhat replied:
string bloom filter will be slow (as it is meant for variable size).  
is it ok ?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,108 @@
+import org.projectnessie.model.IcebergView;
+
+/** A utility class wrapping bloom filter functionality. */
+public class ContentBloomFilter implements Serializable {
```
This class should have a unit test - even if it's rudimentary.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,48 @@
+import org.projectnessie.model.ContentKey;
+
+/** Contains details about one content id: key and the expired values. */
+public class ContentValues implements Serializable {
```
This class requires unit tests.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  void gotValue(Content content, ContentKey key) {
+    synchronized (this) {
```
Synchronziation here is applied in the wrong way. updates are protected, reads are not. That's inconsistent and dangerous. If it's intentional, it should clarify that via code (not just a comment).

ajantha-bhat replied:
removed the synchronization as each reference we add contents sequentially.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GcUtil.java:
```diff
@@ -0,0 +1,74 @@
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.LogResponse;
+
+public class GcUtil {
```
utility class should be `finel` and have a private default c'tor

snazy reviewed (commented):
Looking much better, but I'd really like to see a few more tests - especially around a case when a content-object is expired on one reference and not on another - i.e. that that content is not GC'd at all.

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRest.java:
```diff
@@ -17,8 +17,18 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Please move all the non-GC related changes in AbstractRest hierarchy to a separate PR

snazy commented on model/src/main/java/org/projectnessie/model/Content.java:
```diff
@@ -40,7 +41,7 @@
 @JsonSubTypes({@Type(IcebergTable.class), @Type(DeltaLakeTable.class), @Type(IcebergView.class)})
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
-public abstract class Content {
+public abstract class Content implements Serializable {
```
I wouldn't mind to make all model objects serializable - but please in a separate PR.

ajantha-bhat commented:
> Looking much better, but I'd really like to see a few more tests - especially around a case when a content-object is expired on one reference and not on another - i.e. that that content is not GC'd at all.

@snazy : I will update the testcases , `testMultiRefSharedTable` is already having it for one key, I will add few more keys for the same. 

But I found another bug.
a) scenario 1:
```
create branch1 
t1:42 (expired)
t1:43 (expired)
delete branch1
--- cutoff time--

```

a) scenario 2:
```
create branch1 
t1:42 (expired)
t1:43
--- cutoff time--
delete branch1
```

Scenario 2 is handled, but scenario 1 is not handled. For this to work, I should know whether the ref is dropped before or after cutoff time using reflog. Based on that I can fix it..🤯

ajantha-bhat commented:
>  Scenario 2 is handled, but scenario 1 is not handled. For this to work, I should know whether the ref is dropped before or after cutoff time using reflog. Based on that I can fix it..🤯

@snazy:  Fixed and pushed with test cases as a separate commit for easy review.

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,482 @@
+    // ------  Time ---- | --- branch1 -------------| ---- branch2 -----------   |
+    //         t0        | create branch            |                            |
+    //         t1        |                          | create branch              |
+    //         t2        | TABLE_ONE : 42           |                            |
```
the testcase will have live entires from all the previous ref. Hence only comparing the expected entries. 
Manually verified.

ajantha-bhat replied:
I have passed the involved refs and handled the validation

ajantha-bhat commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/gc/AbstractRestGCTest.java:
```diff
@@ -0,0 +1,482 @@
+    //         t1        |            TABLE_TWO : 42 (expired)
+    //         t2        |            TABLE_TWO : 43
+    //         t3        | ------- cut off time ---------|
+    //         t4        |            TABLE_TWO : 44
```
added

ajantha-bhat commented:
> Looking much better, but I'd really like to see a few more tests - especially around a case when a content-object is expired on one reference and not on another - i.e. that that content is not GC'd at all.

@snazy : PR is ready for review. Replied to all the comments. Added new test cases to cover all the known scenarios.
Improved testcase validation. Added UTs for bloom filter and contentValues.

I think only open discussion is about whether to have a single bloom filter or content type bloom filter. I feel single string bloom filter can be slow compared to content specific bloom filters.  we can even handle this in follow up.
Please let me know.

ajantha-bhat commented:
> I think only open discussion is about whether to have a single bloom filter or content type bloom filter. I feel single string bloom filter can be slow compared to content specific bloom filters. we can even handle this in follow up.
Please let me know.

made a single bloom filter.

snazy reviewed (commented):
Just noticed that all classes are in the packge `org.projectnessie.versioned.persist.gc`, that's not correct. Should be `org.projectnessie.gc.base`.

Few more minor changes that should get in, and the follow-up issue to configure the dead-reference-cutoff-time filed. However, the latter shouldn't be that much work, so I'd prefer to put it into this PR.

The algorithm finally looks correct and the tests seem to cover all situations.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,61 @@
+public class ContentValues implements Serializable {
+
+  private static final long serialVersionUID = -1268447510930923804L;
+  private ContentKey contentKey;
```
I too don't have a clarity at the moment. Once we finalize the consumer (phase2) design. we can change this if it requires. It will be a small change. I thought consumer cannot understand content id. So, I kept the content key.
For example Iceberg's expire snapshots need table name, not the content id. so `contentKey` is needed.

snazy replied:
Again: the table name is only valid at a specific commit.
You're unsure whether and how it will be useful - so it's better to remove it here and add it when it's needed.

ajantha-bhat replied:
I think I need a map of ref + hash  --> key.

I will remove keys now. 

Later while working on ph2 I will add it with other required info.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentValues.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
removed.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/ContentBloomFilter.java:
```diff
@@ -0,0 +1,90 @@
+  // String bloom filter with content type prefix + snapshot/version id.
+  private final BloomFilter<String> icebergContentBloomFilter;
+
+  private static final String ICEBERG_TABLE_PREFIX = "T";
```
I want to save some characters here Instead of long strings from enum name as prefix is used for each content.

snazy replied:
Youi're not saving anything with that

ajantha-bhat replied:
technically bloom filter has to create index based on few characters compared to using enum name (for each content). But I don't want to argue on this. I will change to Enum.name()

snazy replied:
If you care about that, then it would be better to have `Funnel` implementations per content-type.
But anyway - it's already way less than multiple bloom filters easily wasting megabytes of heap.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,379 @@
+              .fetch(FetchOption.ALL)
+              .get()
+              .getMetadata();
+      long totalCommitsInDefaultReference =
```
Calculating this value here is dangerous and will lead to BFs with different sizes that are in turn not mergeable.
The BF sizing parameters must be the same in every spark task.

Best would be to move the code that runs as a separate spark task to a separate class, so it's more obvious what's running where and when it gets initialized.

ajantha-bhat replied:
only public methods (just two) in `DistributedIdentifyContents` happens in spark driver, rest all happens in executor. It is hard to separate. So I think it is ok to keep in same file. 

I now moved `totalCommitsInDefaultReference` calculation to driver and passing this value to executor.

ajantha-bhat commented:
@snazy : changed package name, removed contentkey, added support for dead reference cutoff time along with new testcase `testSingleRefDeadRefCutoff`

ajantha-bhat commented:
@snazy : Finally done !

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,392 @@
+    }
+  }
+
+  private Instant getCutoffTimeForRef(
```
nit: can/should be `static`

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,392 @@
+            config.getOrDefault(
+                GCConfigConstants.CONF_NESSIE_GC_SPARK_TASK_COUNT,
+                String.valueOf(references.size())));
+    if (taskCount <= 0) {
```
if the config's invalid and the implementation silently behaves in an unexpected way, that's surprising for users. better error out early.

ajantha-bhat replied:
moved all conf validation to beginning

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,392 @@
+            .parallelize(references, getTaskCount(references))
+            .map(
+                reference ->
+                    computeLiveContents(
```
Considering the issue w/ the BF siizing I still favor moving the "remote" part to a separate class.
To be clear here: that's for easier maintenance and clear separation of duties - especially for people who are not familiar with the code flow.

ajantha-bhat replied:
done. moved to a separate file

snazy commented:
> @snazy : Finally done !

Sorry to disappoint you, but you missed one point. It's hard to review when things slightly change and changes in such a rather complex PR are squashed together.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,207 @@
+      } catch (NessieNotFoundException ex) {
+        // Ignore the failure and use default value.
+      }
+      Long totalCommitsInDefaultReference =
```
As I am passing from driver to executor, I thought boxed type is better than primitive.

ajantha-bhat replied:
passed primitive now

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,392 @@
+   */
+  public Map<String, ContentBloomFilter> getLiveContentsBloomFilters(
+      List<Reference> references,
+      Map<String, Instant> cutOffTimestampPerRef,
```
In a follow up PR please

ajantha-bhat commented:
@snazy : pushed as a separate commit for easy review

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContentsExecutor.java:
```diff
@@ -0,0 +1,293 @@
+   * @param configuration map of client builder configurations.
+   * @return {@link NessieApiV1} object.
+   */
+  protected static NessieApiV1 getApi(Map<String, String> configuration) {
```
should be moved to GCUtil as it is used from this class and DistributedIdentifyContents

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContentsExecutor.java:
```diff
@@ -0,0 +1,293 @@
+              // then commit is live.
+              commitMeta.getCommitTime().compareTo(cutOffTimestamp) >= 0;
+
+      ImmutableGCParamsExecutor gcParamsExecutor =
```
This one doesn't make sense to me.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,122 @@
+ * Identify the expired and live contents in a distributed way using the spark and bloom filter by
+ * walking all the references (both dead and live).
+ */
+public class DistributedIdentifyContents implements Serializable {
```
Since this class should not be serialized to the executors, the `Serializable` can go away, right?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,266 @@
+        .collect(Collectors.toList());
+  }
+
+  private Instant getInstantFromMicros(Long microsSinceEpoch) {
```
`static` missing

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,266 @@
+    }
+  }
+
+  private List<Reference> collectDeadReferences(
```
`static` missing

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:
```diff
@@ -0,0 +1,122 @@
+            .parallelize(gcParams.getReferences(), getTaskCount(gcParams.getReferences()))
+            .map(
+                reference ->
+                    executor.computeLiveContents(
```
This whole lambda should be in the class that runs on the executor.

snazy reviewed (changes requested):
TBH, the current state of the code is very confusing and hard to follow and reason about.

There are a couple of things that make it very difficult:
* the use of the new params classes - those are only used to wrap parameters for a single method invocation, which is not what those are meant for nor what I had in mind. I asked for a _configuration_ object and an object that contains the immutable discovered _state_.
* a lot of back and forth within the code

I'd suggest to _clearly_ separate the concern of and within the classes.
* GCImpl, not serializable - main entry point - class not to be used in executors
* DistributedIdentifyContents, not serializable - used to tigger work on exectors and collect their results - class not to be used in executors
* DistributedIdentifyContentsExecutor, serializable - code that runs on executors - takes configuration and state in the constructor. the name of this class could be improved to clearly separate it from the latter.

GC-Configuration from Spark configuration should go into GCParams - and validated in GCParams.

ajantha-bhat commented:
> TBH, the current state of the code is very confusing and hard to follow and reason about.

Fixed.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,257 @@
+        : 0;
+  }
+
+  private static void withConfig(
```
Everything in here should be in GCParams. Immutables has built-in support for validation - so please use it.

With that, all the stuff in GCConfigCOnstants can go into GCParams

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,257 @@
+   * @param cutOffTimestampPerRef the cutoff timestamp per ref
+   */
+  public GCImpl(
+      Map<String, String> configuration,
```
Should take only GCParams as a parameter.
The validation is just unnecessary here w/ immutables,

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentBloomFilter.java:
```diff
@@ -0,0 +1,82 @@
+            ? totalCommitsInDefaultReference
+            : gcParams.getBloomFilterExpectedEntries();
+    double fpp = gcParams.getBloomFilterFpp() == null ? 0.03D : gcParams.getBloomFilterFpp();
+    this.icebergContentBloomFilter =
```
Why doesn't this use immutable's mechanism for default values?

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentBloomFilter.java:
```diff
@@ -0,0 +1,82 @@
+
+  public ContentBloomFilter(GCParams gcParams, long totalCommitsInDefaultReference) {
+    long expectedEntries =
+        gcParams.getBloomFilterExpectedEntries() == null
```
This should be handled in GCImpl

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentBloomFilter.java:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  public void put(Content content) {
+    switch (content.getType()) {
```
The switch here is nearly identical to the one in mightContain

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -0,0 +1,284 @@
+            }
+          });
+    } catch (NessieNotFoundException e) {
+      LOGGER.info("Reference {} to retrieve commits no longer exists", reference);
```
This should throw - not ignore

ajantha-bhat commented:
@snazy : Did some self review and pushed the changes again. Please have a look.

snazy commented:
Impossible to review, if it doesn't build.

ajantha-bhat commented:
> Impossible to review, if it doesn't build.

@snazy : Build was passing, then @nastra's PR got merged and gave conflict. I supported for spark3.1 now

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentBloomFilter.java:
```diff
@@ -0,0 +1,75 @@
+
+  public ContentBloomFilter(GCParams gcParams, long totalCommitsInDefaultReference) {
+    long expectedEntries =
+        gcParams.getBloomFilterExpectedEntries() == null
```
This conditional shouldn't be here - it should be in GCImpl.

ajantha-bhat replied:
why not reply to this ?
<img width="838" alt="image" src="https://user-images.githubusercontent.com/5889404/154662691-3c90cc48-f490-426a-93f9-1ff8fd9df582.png">

snazy replied:
YOu can still compute the effective value in GCImpl and not do it here,.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentBloomFilter.java:
```diff
@@ -0,0 +1,75 @@
+            : gcParams.getBloomFilterExpectedEntries();
+    this.icebergContentBloomFilter =
+        BloomFilter.create(
+            Funnels.stringFunnel(Charset.defaultCharset()),
```
What happens, if the default charset is different on separate machines??

ajantha-bhat replied:
ok, I will use StandardCharsets.UTF_8

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,186 @@
+              throw new RuntimeException(
+                  entry.getOperation() + " operation found in dead reflog query");
+          }
+          switch (entry.getRefType()) {
```
I'd prefer to do the switch in `RefUtil` and only do `droppedReferenceTimeMap.put(RefUtil.fromRefLogEntry(entry), getInstantFromMicros(entry.getOperationTime()))` here.

I'm sure that I've mentioned `RefUtil` before

ajantha-bhat replied:
I have replied to that saying it is a new module dependency just for one if check. Is it really need to do ?
but I didn't get reply

snazy replied:
Ah, right - RefUtil isn't available here.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCParams.java:
```diff
@@ -0,0 +1,97 @@
+   * parallelism if not configured.
+   */
+  @Nullable
+  Integer getTaskCount();
```
Is it "task count" or parallelism?

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCUtil.java:
```diff
@@ -0,0 +1,119 @@
+                logEntry -> {
+                  if (!liveCommitPredicate.test(logEntry.getCommitMeta())
+                      && isLiveContentsKeyAdded.isTrue()
+                      && (liveContentKeys.isEmpty())) {
```
As explained in the algorithm, I am collecting the live contents key at cutoff time to retain commit head of those content keys even though they are expired based on cutoff time to support time travel.
I am removing the contents key once commit head is found in expired commits. Hence the `isEmpty()` check and I should not stop traversing before cutoff time hence  `isLiveContentsKeyAdded.isTrue().`

I will try to update the comments.

snazy replied:
It's not the comments - it's the code.
It relies on mutating objects, which are updated in a completely different part of the code base.
It should be a predicate passed into `traverseLiveCommits`

ajantha-bhat commented:
@snazy : Handled all the comments and pushed as new commit for easy review.

tomekl007 commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCParams.java:
```diff
@@ -0,0 +1,94 @@
+   * hours.
+   */
+  @Value.Default
+  default int getCommitProtectionDuration() {
```
are we not able to return `Duration` here? 
If not, I think it should be renamed to `getCommitProtectionHours`.
Now, the caller needs to look into implementation to see that the unit is hours.

snazy replied:
(Mentioned it above) it should be a `Duration`, which would fit well with `Instant`

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCUtil.java:
```diff
@@ -0,0 +1,119 @@
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.LogResponse;
+
+public final class GCUtil {
```
`final` class having member declared 'protected' gets an IDE warning.
So, kept default specifier.

ajantha-bhat replied:
I removed public and so it is (no modifier) package protected.
But still you made comment unresolved. So, I specifically mentioned about `final`

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCParams.java:
```diff
@@ -0,0 +1,97 @@
+   * hours.
+   */
+  @Value.Default
+  default Integer getCommitProtectionTime() {
```
oh, this also I thought renaming comment.

ajantha-bhat replied:
fixed.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentValues.java:
```diff
@@ -0,0 +1,39 @@
+  // Note that we may need more info like refName, hashOnRef, contentKey
+  // for the consumer of GC along with expired contents.
+  // Hence, a separate class.
+  private final Set<Content> expiredContents = new HashSet<>();
```
worst case most of the commits in the reference.
Future we have plan to clean nessie backend and also if GC is run frequently. This won't grow much.

snazy replied:
Here you say "won't grow much" and in the comment below you say "traversing all the commits (imagine millions)". That's quite a difference.

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/ContentBloomFilter.java:
```diff
@@ -0,0 +1,69 @@
+    }
+  }
+
+  private String getValue(Content content) {
```
@snazy : I think it is ok do this. WDYT ?

ajantha-bhat replied:
org.projectnessie.model.Content is our public interface with Nessie clients. So, I am waiting for others opinion

snazy replied:
Tomasz is right here

ajantha-bhat commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:
```diff
@@ -0,0 +1,189 @@
+    } catch (NessieNotFoundException ex) {
+      throw new RuntimeException(ex);
+    }
+    return (defaultRefMetadata != null && defaultRefMetadata.getNumTotalCommits() != null)
```
when the fetch option is not mentioned it is definitely null.
But as it is a @Nullable in immutables declaration, IDE thinks it can be null in my case also.
To avoid warnings added a null check.

snazy replied:
`defaultRefMetadata` should always be non-null here - everything else is an error.

ajantha-bhat replied:
Removed null checks. But IDE errors will confuse the readers. They think it can be null.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -0,0 +1,269 @@
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap) {
+    IdentifiedResult result = new IdentifiedResult();
+    Instant commitProtectionTime =
+        (gcParams.getCommitProtectionDuration() == Duration.ZERO)
```
From the javadoc of `Duration`:
_This is a value-based class; use of identity-sensitive operations (including reference equality (==), identity hash code, or synchronization) on instances of Duration may have unpredictable results and should be avoided. The equals method should be used for comparisons._

I know, that the _current implementation_ makes reference equality for `ZERO` possible, but that does not mean that it will be always the case.

snazy commented on gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:
```diff
@@ -0,0 +1,269 @@
+    IdentifiedResult result = new IdentifiedResult();
+    Instant commitProtectionTime =
+        (gcParams.getCommitProtectionDuration() == Duration.ZERO)
+            ? null
```
The "if protection-duration == 0" condition can be completely avoided. Having a duration of 0 means no protection, which means "now" - so no special case necessary.

ajantha-bhat replied:
removed.

snazy commented:
LGTM now.

Please update the PR description w/ the intentions of this PR and an overall description of the change.

ajantha-bhat commented:
> Please update the PR description w/ the intentions of this PR and an overall description of the change.

@snazy: Updated.

snazy reviewed (approved):
Please use the updated description as the merge-commit comment, not the auto generated one.
